### DISCUSSION TITLE
New APIs - Please feedback on rewriting FlowStacks

### DISCUSSION THREAD

**Discussion by johnpatrickmorgan** [johnpatrickmorgan/FlowStacks#51] [Announcements]
I've been experimenting with new FlowStacks APIs based on the latest SwiftUI `NavigationStack` APIs. I've rewritten the library with essentially the same entities, but with `Navigation` replaced with `Flow`: 

🔀 NavigationStack -> FlowStack

🔀 NavigationLink -> FlowLink

🔀 NavigationPath -> FlowPath

🔀 navigationDestination -> flowDestination

🔀 NavigationPath.CodableRepresentation -> FlowPath.CodableRepresentation

The `FlowStacks` APIs expand the `NavigationStack` APIs:

- As well as managing navigation, you can manage sheet and full-screen cover presentation. 
- A `FlowNavigator` environment object allows you to easily push, pop, present etc. from deep in your view hierarchy.
- The flowDestination APIs also make the screen's data available as a Binding.
- Not limited to just the latest versions of SwiftUI.

These APIs offer advantages in decoupling the building of screens from managing flow state, so you can more easily break up navigation across different files or modules, and should feel familiar to SwiftUI users.

The main disadvantage I see is that you can no longer change the root screen via the routes array or path. However, I think that leads to better modelling of app flows. 

[Here's the branch](https://github.com/johnpatrickmorgan/FlowStacks/tree/new-api) if you'd like to try it out - I'd like to hear opinions on this rewrite before committing to it as it's a significant change. Thanks!

**Comment by johnpatrickmorgan** [comment]
Here's an example of rewriting code that uses the current FlowStacks APIs:

<details>
 <summary>Current API</summary>

```swift
enum Screen {
  case home
  case numberList
  case numberDetail(Int)
}

struct AppCoordinator: View {
  @State var routes: Routes<Screen> = [.root(.home)]
    
  var body: some View {
    Router($routes) { screen, _ in
      switch screen {
      case .home:
        HomeView()
      case .numberList:
        NumberListView()
      case .numberDetail(let number):
        NumberDetailView(number: number)
      }
    }
  }
}
```

</details>

<details>
 <summary>Proposed API</summary>

```swift
enum Screen {
  case numberList
  case numberDetail(Int)
}

struct AppCoordinator: View {
  @State var routes: Routes<Screen> = []
    
  var body: some View {
    FlowStack(path: $routes) { 
      HomeView()
        .flowDestination(for: Screen.self) { screen in
          switch screen {
          case .numberList:
            NumberListView()
          case .numberDetail(let number):
            NumberDetailView(number: number)
        }
    }
  }
}
```

</details>

There's no longer a type guarantee that the routes array's screen type matches the destination builder's screen argument, but that's what allows the destination builder to be decoupled from the state management. The root screen is no longer part of the routes array, which might be awkward if your flow required you to swap out the root screen. In those cases, I imagine you would need to either:
- Have a root screen that conditionally shows one root or the other, OR
- Split the flow into two separate flows, each with a FlowStack and a parent view that switches between them as needed.

**Comment by DavidKmn** [comment]
Great work @johnpatrickmorgan, playing around on the new branch this week. The plan is to add this new api and slowly deprecate the other api?

**Reply by DavidKmn** [reply]
Also I suspect this fixes #47 ?

**Reply by johnpatrickmorgan** [reply]
> The plan is to add this new api and slowly deprecate the other api?

Yes, assuming people are happy with the new API. 

> Also I suspect this fixes https://github.com/johnpatrickmorgan/FlowStacks/pull/47 ?

Sadly, I don't believe it does, as we're still using the same `NavigationView` API under the hood.

**Reply by DavidKmn** [reply]
Sad, that warning is really annoying, have you found a way to maybe mute it?

**Comment by DavidKmn** [comment]
Also think an example with a View Model (ObservableObject) would be a good to have in the discussion

**Reply by johnpatrickmorgan** [reply]
Good idea, I'll try to add that.

**Reply by johnpatrickmorgan** [reply]
I've (finally!) added [an example](https://github.com/johnpatrickmorgan/FlowStacks/blob/new-api/Example/Example/Shared/NumberVMFlow.swift) now.

**Comment by DavidKmn** [comment]
Speaking no the main disadvantage of not being able to start off with the root through the routes array, how do you propose we go about a case where the root view needs a view model to be initialised?

**Reply by johnpatrickmorgan** [reply]
I'll try and include that in the example, but I assume the root view model would be stored in the same place as the Routes array / FlowPath.

**Reply by johnpatrickmorgan** [reply]
Here's an [example](https://github.com/johnpatrickmorgan/FlowStacks/blob/new-api/Example/Example/Shared/NumberVMFlow.swift).

**Comment by savage7** [comment]
I like the new api :blush: @johnpatrickmorgan, great work! I don't see a major disadvantage with the missing root navigation, basically thats the same  with now having a path or specific flow for tab based navigation. 

How are subflows/coordinators handled with this approach? I didn't look at the code but I guess that every FlowStack has its FlowPathNavigator or are they shared when a FlowStack is used within another FlowStack?

**Reply by johnpatrickmorgan** [reply]
Thanks @savage7 - I'm currently looking into the best way to allow composition of FlowStacks, and I'm exploring the possibility that child FlowStacks might defer their state management to their parent FlowStack. This should work well if the child FlowStack wasn't created with its own state binding, but might have to be treated as an error if the child has an Array binding. And perhaps the outer and inner state could be synced if the child has its own FlowPath binding. 

<table class="tg">
<thead>
  <tr>
    <th class="tg-73oq"></th>
    <th class="tg-73oq"></th>
    <th class="tg-0pky" colspan="3">Inner FlowStack Binding</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-73oq"></td>
    <td class="tg-73oq"></td>
    <td class="tg-0pky">NONE</td>
    <td class="tg-0pky">PATH</td>
    <td class="tg-0pky">ARRAY</td>
  </tr>
  <tr>
    <td class="tg-0pky" rowspan="3"><span style="font-weight:400;font-style:normal;text-decoration:none">Outer FlowStack Binding</span></td>
    <td class="tg-0pky">N</td>
    <td class="tg-0pky">✅</td>
    <td class="tg-0pky">Sync with outer?</td>
    <td class="tg-0pky">🚫</td>
  </tr>
  <tr>
    <td class="tg-0pky">P</td>
    <td class="tg-0pky"><span style="font-weight:400;font-style:normal;text-decoration:none">✅</span></td>
    <td class="tg-0pky">Sync with outer?</td>
    <td class="tg-0pky"><span style="font-weight:400;font-style:normal;text-decoration:none">🚫</span></td>
  </tr>
  <tr>
    <td class="tg-0lax">A</td>
    <td class="tg-0lax"><span style="font-weight:400;font-style:normal;text-decoration:none">Add utilities to transform inner types?</span></td>
    <td class="tg-0lax">🚫</td>
    <td class="tg-0lax"><span style="font-weight:400;font-style:normal;text-decoration:none">🚫 or sync if same type?</span></td>
  </tr>
</tbody>
</table>

**Comment by mbuchetics** [comment]
@johnpatrickmorgan what's the current state here? I am thinking about using FlowStacks in a new project and I am wondering whether I should use the latest released version or the new-api branch?

**Reply by johnpatrickmorgan** [reply]
Hi @mbuchetics! I expect the library to adopt the proposed API at some point, but I still have outstanding tasks (documentation, tests, work on composition of child FlowStacks discussed above). I haven't had the time to dedicate to these tasks recently so I don't expect to release the new API for a while still. I would recommend you use the current released version. It shouldn't be too painful to migrate to the new API when it's eventually ready, though I would recommend you avoid using a single Router if it means you need to swap out the root screen. In that case, I would recommend having multiple Routers, one for each root screen.

**Comment by Aduci01** [comment]
Dear @johnpatrickmorgan , thank you for the great framework. Since this rewriting experiment is over 2 years old, may I ask you if you still consider finishing it?

**Reply by johnpatrickmorgan** [reply]
Sorry for the delay - it's now released as v0.6.0.

**Comment by johnpatrickmorgan** [comment]
Thanks for your patience - this rewrite has been a long time in the works but has now been released as v0.6.0. Please try it out and let me know if you encounter any issues. If you want to migrate an existing project, please see the [migration docs](https://github.com/johnpatrickmorgan/FlowStacks/blob/main/Docs/Migration/Migrating%20to%201.0.md). Thank you!
